third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor_v2.rs

```diff
@@ -0,0 +1,1883 @@
+                self.cur_loc(),
+                format!(
+                    "cannot extract resource `{}` which is still borrowed",
+                    self.global_env().display(resource)
+                ),
+                "extracted here",
+                self.borrow_info(&label, |_| true).into_iter(),
+            )
+        }
+        self.state.moved.remove(&dest);
+    }
+
+    /// Process a return instruction.
+    fn return_(&mut self, srcs: &[TempIndex]) {
+        for src in srcs {
+            if self.ty(*src).is_reference() {
+                // Need to check whether this reference is derived from a local which is not a
+                // a parameter
+                if let Some(label) = self.state.label_for_local(*src) {
+                    for root in self.state.roots(label) {
+                        for location in self.state.node(&root).locations.iter() {
+                            match location {
+                                MemoryLocation::Global(resource) => self.error_with_hints(
+                                    self.cur_loc(),
+                                    format!(
+                                        "cannot return a reference derived from global `{}`",
+                                        self.global_env().display(resource)
+                                    ),
+                                    "returned here",
+                                    self.borrow_info(&root, |_| true).into_iter(),
+                                ),
+                                MemoryLocation::Local(local) => {
+                                    if *local >= self.target().get_parameter_count() {
+                                        self.error_with_hints(
+                                            self.cur_loc(),
+                                            format!(
+                                                "cannot return a reference derived from {} since it is not a parameter",
+                                                self.display(*local)
+                                            ),
+                                            "returned here",
+                                            self.borrow_info(&root, |_| true).into_iter(),
+                                        )
+                                    }
+                                },
+                                MemoryLocation::External | MemoryLocation::Derived => {},
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        self.state.moved.extend(srcs.iter().cloned())
+    }
+
+    /// Process a ReadRef instruction. In contrast to `self.check_read_local`, this needs
+    /// to check the value behind the reference.
+    fn read_ref(&mut self, dest: TempIndex, src: TempIndex) {
+        debug_assert!(self.is_ref(src));
+        self.check_write_local(dest);
+        self.check_read_local(src, ReadMode::Argument);
```

Comment:
This will always return without doing any checks because of the check at the beginning of `check_read_local` (we already know from the debug assert above that `src` is a ref)?